src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -11799,18 +11887,19 @@ int Assembler::vex_prefix_and_encode(int dst_enc, int nds_enc, int src_enc, VexS
             (!_legacy_mode_vl) ||
             (attributes->is_legacy_mode())),"XMM register should be 0-15");
     // Instruction with legacy_mode true should have dst, nds and src < 15
-    assert(((dst_enc < 16 && nds_enc < 16 && src_enc < 16) || (!attributes->is_legacy_mode())),"XMM register should be 0-15");
+    assert(((!is_extended) || (!attributes->is_legacy_mode())),"XMM register should be 0-15");
   }
 
   clear_managed();
   if (UseAVX > 2 && !attributes->is_legacy_mode())
   {
     bool evex_r = (dst_enc >= 16);
     bool evex_v = (nds_enc >= 16);
+    bool evex_b = (src_enc >= 16) && src_is_gpr;
     // can use vex_x as bank extender on rm encoding
-    vex_x = (src_enc >= 16);
+    vex_x = (src_enc >= 16) && !src_is_gpr;
     attributes->set_is_evex_instruction();
-    evex_prefix(vex_r, vex_b, vex_x, evex_r, evex_v, nds_enc, pre, opc);
+    evex_prefix(vex_r, vex_b, vex_x, evex_r, evex_b, evex_v, false /*eevex_x*/, nds_enc, pre, opc);
```

Comment:
Thanks @jatin-bhateja .  Do you mean a check such as:

`if ((UseAVX > 2 || UseAPX) && !attributes->is_legacy_mode())` ?

Reply:
> UseAPX

Yes, attaching a test depicting incorrectness with UseAVX=2 for SHLX which is a legacy map 2 instruction promotable to extended EVEX with EGPR operands.
[shift_left_APX.txt](https://github.com/openjdk/jdk/files/15261495/shift_left_APX.txt)

It will not be appropriate to modify VM_Version::supports_evex for APX feature since its used for constraining dynamic register classes associated with vector operands.

Reply:
Ok, thanks.  I've added the above change to the conditionals in the vex_prefix and vex_prefix_and_encode functions.